File: lib/to_mdx/parsers/open_document/text/plain_text.rb
# frozen_string_literal: true

require_relative "../base"

module ToMdx
  module Parsers
    module Text
      class PlainText < Base
        attr_accessor :text

        def initialize(text_node)
          super(text_node)
          @text = decode_html_entities(text_node.to_s)
        end

        def empty?
          @text.empty?
        end

        def to_mdx
          Base.escape_for_mdx(@text)
        end

        def to_s
          @text
        end

        private

        def decode_html_entities(text)
          text.gsub(/&quot;/, '"')
               .gsub(/&lt;/, '<')
               .gsub(/&gt;/, '>')
               .gsub(/&amp;/, '&')
        end
      end
    end
  end
end

File: src/parsers/open-document/text/plain_text.ts
import { escapeForMdx } from "../../../utils.ts";
import { type ParseContext, type ParsedElement } from "../../base_element.ts";

export class PlainText implements ParsedElement {
  private text: string;

  readonly childElementCount = 0;

  constructor(text: string) {
    this.text = decodeHtmlEntities(text);
  }

  toMdx(): string {
    return escapeForMdx(this.text);
  }

  toString(): string {
    return this.text;
  }

  empty(): boolean {
    return this.text.length === 0;
  }

  attr(_name: string): string | null {
    return null;
  }

  context<K extends keyof ParseContext>(_key: K): ParseContext[K] {
    return null;
  }

  [Symbol.iterator](): Iterator<ParsedElement> {
    return {
      next() {
        return { value: undefined, done: true };
      },
    };
  }
}

const decodeHtmlEntities = (text: string): string => {
  return text
    .replaceAll(/&quot;/g, '"')
    .replaceAll(/&lt;/g, "<")
    .replaceAll(/&gt;/g, ">")
    .replaceAll(/&amp;/g, "&");
};
